packages/actor-abstract-path/lib/ActorAbstractPath.ts

```diff
@@ -84,4 +108,77 @@ export abstract class ActorAbstractPath extends ActorQueryOperationTypedMediated
       }
     });
   }
+
+  // Let the iterator `it` emit all bindings of size 2, with subjectStringVariable as value subjectVal
+  // and objectStringVariable as value all nodes reachable through predicate* beginning at objectVal
+  public async ALPTwoVariables(subjectString: string, objectString: string, subjectVal: Term,
+    objectVal: Term, predicate: Algebra.PropertyPathSymbol, context: ActionContext,
+    termHashesGlobal: {[id: string]: Promise<Term[]>}, termHashesCurrentSubject: {[id: string]: boolean},
+    it: BufferedIterator<Bindings>, counter: any): Promise<void> {
+    const termString = termToString(objectVal);
+
+    // If this combination of subject and object already done, return nothing
+    if (termHashesCurrentSubject[termString]) {
+      return;
+    }
+
+    termHashesCurrentSubject[termString] = true;
+
+    (<any> it)._push(Bindings({ [subjectString]: subjectVal, [objectString]: objectVal }));
+    counter.count++;
+
+    // If every reachable node from object has already been calculated, use these for current subject too
+    if (termString in termHashesGlobal) {
+      const objects = await termHashesGlobal[termString];
+      for (const object of objects) {
+        if (!termHashesCurrentSubject[termToString(object)]) {
+          termHashesCurrentSubject[termToString(object)] = true;
+          (<any> it)._push(Bindings({ [subjectString]: subjectVal, [objectString]: object }));
+        }
+      }
+      if (--counter.count === 0) {
+        it.close();
+      }
+
+      return;
+    }
+
+    // Construct promise to calculate all reachable nodes from this object
+    const promise = new Promise<Term[]>(async(resolve, reject) => {
+      const objectsArray: Term[] = [];
+
+      const thisVariable = this.generateVariable();
+      const vString = termToString(thisVariable);
+      const path = ActorAbstractPath.FACTORY.createPath(objectVal, predicate, thisVariable);
+      const results = ActorQueryOperation.getSafeBindings(
+        await this.mediatorQueryOperation.mediate({ operation: path, context }),
+      );
+
+      results.bindingsStream.on('data', async bindings => {
+        const result = bindings.get(vString);
+        objectsArray.push(result);
+        await this.ALPTwoVariables(
+          subjectString,
+          objectString,
+          subjectVal,
+          result,
+          predicate,
+          context,
+          termHashesGlobal,
+          termHashesCurrentSubject,
+          it,
+          counter,
+        );
+      });
+      results.bindingsStream.on('end', () => {
+        if (--counter.count === 0) {
+          it.close();
+        }
+        resolve(objectsArray);
+      });
+    });
+
+    // Set it in the termHashesGlobal when this object occurs again they can wait for this promise
+    termHashesGlobal[termString] = promise;
```

Comment:
Just realized this, but should there not be a `return promise;` below this line? Because currently the promise gets generated, but if the `if` statement above is not entered (which can happen based on the data structure), the promise above could potentially never be explicitly executed.

Reply:
This is what I found on https://medium.com/@adityaprakashjoshi1/promises-in-javascript-cf81f8e56f80 : The executorfunction is executed immediately by the Promise implementation, passing resolve and reject functions (the executor is called before the Promise constructor even returns the created object).
Return is actually not possible because the return type of the function is Promise<void>.

Reply:
Indeed, promises are started immediately upon creation.